qiskit/circuit/singleton.py

```diff
@@ -0,0 +1,250 @@
+#    that has the immutable overrides in its method-resolution order in the correct place. These
+#    override the standard methods / properties that are defined on the mutable gate (we do not
+#    attempt to override any cases where the type object we are creating has extra inplace methods).
+#
+# 3. We can't instantiate this new subclass, because when it calls `XGate.__init__`, it will attempt
+#    to set some attributes, and these will be rejected by immutability.  Instead, we first create a
+#    completely regular `XGate` instance, and then we dynamically change its type to the singleton
+#    class, freezing it.
+#
+# We could do this entirely within the metaclass machinery, but that would require `XGate` to be
+# defined as something like
+#
+#   class XGate(Gate, metaclass=_SingletonMeta, overrides=_SingletonGateOverrides): ...
+#
+# which is super inconvenient (or we'd have to have `_SingletonMeta` do a bunch of fragile
+# introspection).  Instead, we use the `abc.ABC`/`abc.ABCMeta` pattern of defining a concrete middle
+# class (`SingletonGate` in the `XGate` case) that sets the metaclass, selects the overrides to be
+# applied, and has an `__init_subclass__` that applies the singleton-subclass-creation steps above.
+# The overrides are in separate classes so that _mutable_ `XGate` instances do not have them in
+# their own MROs; doing this is easier to implement, but requires all the setters and checkers to
+# dance around at runtime trying to validate whether mutating the instance is allowed.
+
+
+def _impl_new(cls, *_args, **_kwargs):
+    # __new__ for the singleton instances.
+    raise TypeError(f"cannot create '{cls.__name__}' instances")
+
+
+def _impl_init_subclass(base, overrides):
+    # __init_subclass__ for the classes that make their children singletons (e.g. `SingletonGate`)
+
+    def __init_subclass__(cls, *, create_singleton=True, **kwargs):
+        super(base, cls).__init_subclass__(**kwargs)
+        if not create_singleton:
+            return
+
+        # We need to make a new type object that pokes in the overrides into the correct
+        # place in the method-resolution order.
+        singleton_class = _SingletonMeta.__new__(
+            _SingletonMeta,
+            f"_Singleton{cls.__name__}",
+            (cls, overrides),
+            # This is a dynamically generated class so it's got no module.  The slot layout of the
+            # singleton class needs to match any layout in the base.
+            {"__module__": None, "__slots__": (), "__new__": _impl_new, "_base_class": cls},
+            create_singleton=False,
+        )
+
+        # Make a mutable instance, fully instantiate all lazy properties, then freeze it.
+        cls._singleton_instance = cls(_force_mutable=True)
+        cls._singleton_instance._define()
+        cls._singleton_instance.__class__ = singleton_class
+
+    return classmethod(__init_subclass__)
```

Comment:
I wish proc macros were a thing in python, this kind of structure could be done very simply in rust.

Reply:
tbf, this is kind of the direct equivalent of a auto-derive macro, and I'm not sure the proc macro version of this would have been any simpler lol